src/main/java/org/spongepowered/common/service/sql/SqlServiceImpl.java

```diff
@@ -268,7 +268,11 @@ public static ConnectionInfo fromUrl(@Nullable PluginContainer container, String
             }
             final String unauthedUrl = "jdbc:" + protocol + (hasSlashes ? "://" : ":") + serverDatabaseSpecifier;
             final String driverClass = DriverManager.getDriver(unauthedUrl).getClass().getCanonicalName();
-            return new ConnectionInfo(user, pass, driverClass, unauthedUrl, fullUrl);
+            if (protocol.equals("mysql")) {
+                return new ConnectionInfo(user, pass, driverClass, unauthedUrl + "?disableMariaDbDriver", fullUrl);
```

Comment:
This hack produce invalid result when jdbc url already contains some parameters.

```java
String jdbcUrl = "jdbc:mysql://zml:totallymypassword@localhost/sponge?useSSL=false";
SqlServiceImpl.ConnectionInfo subject = SqlServiceImpl.ConnectionInfo.fromUrl(null, jdbcUrl);
System.out.println(subject.getAuthlessUrl());
// outputs: jdbc:mysql://localhost/sponge?useSSL=false?disableMariaDbDriver
```

Checking for question mark and appending `&disableMariaDbDriver` is not a solution. Possibility of setting custom data source properties by user as key-value map in sponge configs needed. Also there is need to maintain backward compatibility with already existing configurations which put properties in jdbc url. This involves parsing of query string in fromUrl factory method as well.

Reply:
The following unit test 

```
@Test
    public void testMysqlConnectionInfoAdvanced() throws SQLException {
        final String jdbcUrl = "jdbc:mysql://zml:totallymypassword@localhost/sponge?useSSL=false";
        final SqlServiceImpl.ConnectionInfo subject = SqlServiceImpl.ConnectionInfo.fromUrl(null, jdbcUrl);

        assertEquals("zml", subject.getUser());
        assertEquals("totallymypassword", subject.getPassword());
        assertEquals("jdbc:mysql://localhost/sponge?useSSL=false", subject.getAuthlessUrl());
        assertEquals("org.mariadb.jdbc.Driver", subject.getDriverClassName());
    }
```
Shows that (without this PR) the url parameters are just passed straight through.

> Checking for question mark and appending &disableMariaDbDriver is not a solution. Possibility of setting custom data source properties by user as key-value map in sponge configs needed. Also there is need to maintain backward compatibility with already existing configurations which put properties in jdbc url. This involves parsing of query string in fromUrl factory method as well.

I don't think this is strictly necessary for this PR. Is there a good reason why we should explode those properties in the Sponge config, rather then just following the existing standard of jdbc urls?

Alternate things that I had discussed with @brjcarbone in voice, was parsing and manipulating the URL using existing URL building / manipulation. But I havn't had time yet to investigate if there is something simple in the standard libraries, or any of the utility libraries that we use in the Sponge implementation.

If someone has a reliable fix for this, I'm all ears.

I hadn't yet commented on the string concatenation yet, because I only just tested that the parameters indeed do pass through the regex.

Before this I was told the regex would need updating, and that the parameters were stripped.

Reply:
You're actually correct, I misinterpreted the regex and everything after `@` is Group 5. I'll update the regex to consider a sixth group that properly appends options.

Reply:
Yeah, I'm not sure which would be preferred.

Letting the user handle it gives more control, but complicates it for a first time user, meaning documentation etc, which is one of the arguments for accepting this PR in the first place.

Manipulating the URL ourselves, wouldn't be too hard, but manipulating complicated schemes such as emails and URL's has bitten me in the ass before without using a well tested library to do so, If we have something in the classpath that can do that, It'd be a nobrainer to me to add it for mysql.

Reply:
All sponge supported databases accept login and password inside jdbc url.

H2:

```
jdbc:h2:<url>[;USER=<username>][;PASSWORD=<value>]
jdbc:h2:file:~/sample;USER=sa;PASSWORD=123
```

MySQL/MariaDB:

```
jdbc:mysql://localhost:3306/DB?user=root&password=myPassword
```

Sqlite doesn't have any athentication, so there in no syntax for passing credentials.

So, any jdbc url parsing is not needed at all. Introducing custom auth syntax in jdbc url is sponge's major flaw, which must be supported forever. This does not mean that you should touch it too.

> I don't think this is strictly necessary for this PR. Is there a good reason why we should explode those properties in the Sponge config, rather then just following the existing standard of jdbc urls?

You think right. This is not necessary. I just forgot that options in jdbc url may be intermixed with `setDataSourceProperties` method. Hence there is no reasons to introduce new syntax construct in sponge configs and change `SqlServiceImpl` or `ConnectionInfo` signature.

> Letting the user handle it gives more control, but complicates it for a first time user, meaning documentation etc, which is one of the arguments for accepting this PR in the first place.

Exposing this implementation detail to user makes impossible to get rid of mysql connector in the future, when mariadb connector finally implements this new auth scheme. 

The only way to properly inject `disableMariaDbDriver` without affecting user expirience - is adding it in `PROTOCOL_SPECIFIC_PROPS` collection. Need additional testing to ensure that other supported jdbc libraries ignores this option if they do not understand it.